feat: add secretscan attestor to detect secrets in products

colek42 opened the pull request:
# Add SecretScan Attestor

## Summary
This PR adds a new SecretScan attestor that scans products and attestations for leaked secrets, credentials, and sensitive information:

- **Comprehensive Secret Detection**: Uses Gitleaks detection engine to identify a wide range of secrets (API keys, tokens, credentials, etc.)
- **Security-First Design**: Implements multiple security measures to prevent leaking detected secrets:
  - SHA256 hashing of all secret values
  - Truncated prefixes for context without revealing secrets
  - Deterministic finding IDs for cross-referencing
  - Proper source identification with meaningful paths
  
- **Attestation Sanitization**: Automatically sanitizes attestation outputs to prevent secret propagation
- **Configurable Behavior**: Supports various options for customizing detection:
  - Allowlists with regex patterns, path patterns, and stop words
  - Max file size limitations
  - Configurable failure modes

## Features
- **Secure Finding Storage**: All detected secrets are obfuscated with only hash values stored
- **Resource Protection**: Implements size limiting and file type filtering
- **Comprehensive Allowlist**: Support for excluding specific patterns to reduce false positives
- **Meaningful Identifiers**: Sources properly identified as 'attestation:name' or 'product:/path'
- **Proper Error Handling**: Robust error handling throughout the codebase
- **High Test Coverage**: Comprehensive test suite with over 65% coverage

## How It Works
The SecretScan attestor runs after product attestors to scan all products for secrets. It also scans the outputs of other attestors to detect secrets that might be accidentally included in attestations. When secrets are found, they are logged with secure identifiers, and the attestor can either continue or fail based on configuration.

## Security Considerations
- No actual secrets are stored in serialized attestation output
- Temporary files use restricted permissions
- Secret values are properly obfuscated in all outputs
- Binary and very large files are automatically skipped

## Testing
- Comprehensive unit tests
- Table-driven tests for various detection scenarios
- Integration tests to verify end-to-end flows
- Linter compliant with proper error handling

mikhailswift commented on attestation/secretscan/secretscan.go:
```diff
@@ -0,0 +1,449 @@
+	return os.WriteFile(outputPath, []byte(strings.Join(stringsList, "\n")), 0644)
+}
+
+func (a *Attestor) MarshalJSON() ([]byte, error) {
```
This function seems unnecessary

mikhailswift commented on attestation/secretscan/secretscan.go:
```diff
@@ -0,0 +1,449 @@
+	})
+}
+
+func (a *Attestor) UnmarshalJSON(data []byte) error {
```
Similar to above, this doesn't seem to add anything that the default implementations of Marshal/Unmarshal would give us

mikhailswift commented on attestation/secretscan/secretscan.go:
```diff
@@ -0,0 +1,449 @@
+		}
+
+		// Scan the file with Gitleaks
+		findings, err := a.ScanFile(filename, detector)
```
I'm worried calling `ScanFile` for every file in the product list may be incredibly slow. Was using `[DetectDirectory](https://github.com/gitleaks/gitleaks/blob/af7d5bcb83255283767c45e03c904af9a77e6e75/detect/directory.go#L20)` not an option here? It looks like it uses some concurrency, and while it may still be slow it may not be as slow as doing all section in serial.

colek42 replied:
Me too.  I'll remove it for now

jkjell commented on attestation/secretscan/secretscan.go:
```diff
@@ -0,0 +1,449 @@
+
+const (
+	Name    = "secretscan"
+	Type    = "https://witness.dev/attestations/secretscan/v0.1"
```
There's going to be a race condition here with other post product attestors. If a SLSA attestor or SBOM detector have secrets, they may or may not be detected by this.

colek42 replied:
lets not run it against the post product attestors, we are already scanning the products

colek42 commented:
I've implemented the requested changes:\n\n1. Completely removed binary scanning capability as requested\n   - Removed all binary-related code (extractStringsFromBinary, etc.)\n   - Modified the scanProducts method to skip binary files rather than process them\n   - Made product scanning always run (removed conditional scanning)\n\n2. Removed the custom MarshalJSON and UnmarshalJSON methods since they don't add anything beyond the default implementations\n\nThese changes simplify the code significantly while maintaining the core secret scanning functionality. Regarding the performance concerns with scanning files serially, a future enhancement could implement DetectFiles method for better concurrency.

colek42 commented:
# SecretScan Attestor Refactoring - Testing Summary

## Enhancements Made
1. Changed the `Finding` struct to use `cryptoutil.DigestSet` instead of raw strings for secret representation
2. Added detection of hardcoded environment variable names from the sensitive environment variables list
3. Added proper context handling to use consistent hash algorithms across the codebase

## Testing Results

We thoroughly tested the refactored secretscan attestor using the witness CLI and confirmed the following functionality:

1. **Detection of Actual Secrets (via Gitleaks)** - The attestor correctly detects actual secrets using Gitleaks rules and stores them as DigestSet hashes
2. **Detection of Environment Variable Names** - The attestor now detects hardcoded references to sensitive environment variable names
3. **Proper Hash Algorithms** - The attestor uses the hash algorithms from the attestation context consistently

## Conclusion

The refactored secretscan attestor successfully:

1. ✅ Uses DigestSet instead of raw strings for secret representation
2. ✅ Detects sensitive environment variable names in attestations and files
3. ✅ Maintains compatibility with existing witness CLI commands
4. ✅ Includes proper context to use consistent hash algorithms

All tests pass, including the new test cases for environment variable detection, and the functionality has been verified with multiple real-world examples.
EOT < /dev/null

colek42 commented:
# Testing the Enhanced Environment Variable Detection

I've tested the enhanced environment variable detection in the secretscan attestor. Here's a simple test case that demonstrates how it works:

## How to Test

1. Set a sensitive environment variable:
```bash
export AWS_ACCESS_KEY_ID="AKIAMYSPECIALSECRETVALUE"
```

2. Run witness with both environment and secretscan attestors, using the environment variable in the command:
```bash
./bin/witness run --attestations environment,secretscan -o test-env-detection.json --step test-env-vars --signer-file-key-path ./test/testkey.pem -- echo "Hello with AWS key: $AWS_ACCESS_KEY_ID"
```

3. Extract the findings from the attestation:
```bash
jq -r '.payload' test-env-detection.json | base64 -d | jq '.predicate.attestations[4].attestation.findings'
```

## Test Results

The secretscan attestor successfully detects:

1. The sensitive environment variable name in the environment attestation:
```json
{
  "ruleId": "witness-env-var-aws-access-key-id",
  "description": "Sensitive environment variable name: AWS_ACCESS_KEY_ID",
  "location": "attestation:environment",
  "startLine": 6,
  "secret": {
    "sha256": "a7005f9a916d1efab429b5cecb79f518fc3217c023748841281741f05835c1be"
  },
  "match": "\": {\n    \"AWS_ACCESS_KEY_ID\": \"******"
}
```

2. The actual environment variable value in the command-run attestation:
```json
{
  "ruleId": "witness-env-value-aws-access-key-id",
  "description": "Sensitive environment variable value detected: AWS_ACCESS_KEY_ID",
  "location": "attestation:command-run",
  "startLine": 4,
  "secret": {
    "sha256": "f6d1c5b0c20cb9ac6806b9f7bb1bf6d2058c4b0732fd9094fa0bc64c85f80cfd"
  },
  "match": "key: [SENSITIVE-VALUE]\"\n  ]"
}
```

The command-run attestation shows the leaked value:
```json
{
  "cmd": [
    "echo",
    "Hello with AWS key: AKIAMYSPECIALSECRETVALUE"
  ],
  "stdout": "Hello with AWS key: AKIAMYSPECIALSECRETVALUE\n",
  "exitcode": 0
}
```

## Key Advantages

1. **Security**: The actual secret value is never stored in plaintext in the finding - only as a secure DigestSet
2. **Comprehensive Detection**: Detects both hardcoded environment variable names and leaked environment variable values
3. **Redaction**: The match field redacts the actual value with a placeholder
4. **DigestSet Integration**: Uses the same hash algorithms from the attestation context for consistent security

This feature helps detect both explicit references to environment variable names and cases where secret values are inadvertently leaked into command output or other attestations.
